planner/core/logical_plan_builder.go

```diff
@@ -1384,8 +1385,67 @@ func (b *PlanBuilder) buildProjection4Union(ctx context.Context, u *LogicalUnion
 	}
 	u.schema = expression.NewSchema(unionCols...)
 	u.names = names
-	// Process each child and add a projection above original child.
-	// So the schema of `UnionAll` can be the same with its children's.
+
+	allExprs := make([][]expression.Expression, 0)
```

Comment:
why not use https://github.com/pingcap/tidb/pull/27212#issuecomment-902726520, we collect all the expressions and do the cast if needed, I think they are same